src/metrics.ts

```diff
@@ -0,0 +1,182 @@
+import { IncomingMessage } from 'http';
+import { Http2ServerRequest } from 'http2';
+import { resolveHostFromRequest, resolveIpFromRequest } from './util';
+
+export type MetricsOptions = {
+  historySize?: number;
+  waitForHistory?: boolean;
+  hostBadActorSplit?: number;
+  ipBadActorSplit?: number;
+  hostWhitelist?: Set<string>;
+  ipWhitelist?: Set<string>;
+}
+
+export class Metrics {
```

Comment:
could timestamps be helpful? to help track snapshots and differentiate between spikes and slow buildups

Reply:
Not with the current algorithm since it uses a sliding window of time. But future versions might use a progressive algorithm and have need for time. We might even be able to move away from N requests and use a time window instead, but out of scope for this PR.